## All rules should be executed in a shell with the environment that you used with Prof. Molineris loaded:
# $ conda

# Bioinfo lessons:
# https://docs.google.com/presentation/d/1bfdVoH5QiCijTPSb-H2zYP0METFRGbfHepQ5zJytpDY/edit#slide=id.g98431aadc5_0_1103
# https://docs.google.com/presentation/d/1K9Y-EuirRfEBMErjx8WgKTe8LlI0decL-KsORn4DAwc/edit#slide=id.gf7c04f0c33_0_15

SAMPLES = ['a','b','c','d','e']

INDEX_DIR='/home/imolineris/biotech4neuron_lessons/dataset/ucsc/dm6_STAR_index'
GTF='/home/imolineris/biotech4neuron_lessons/dataset/ucsc/ENCODE.gtf'
BED='/home/imolineris/biotech4neuro_lessons/dataset/ucsc/dm6_ensembl.bed'

MAP='/home/elenagrassi/biotech4neuro/material/19/map_droso.tsv'

# ...

########################  QC 


rule all_qc:
# ...

rule fastqc:
# ...

######################## Align and count

rule align:
# ...

rule featureCount:
# ...

rule all_counts:
    # ...
    script: "merge_counts.R"

######################## Find expressed genes and produce bed for them

rule extract_expressed_genes:
    # ...
    
rule translate_to_ensembl:
    input: counts="expressed_genes_counts.tsv", map=MAP
    # ...
    
rule bed_expressed_genes:
    # ...
    
######################## Call geneBody_covereage for sample a and e

# ...

######################## Find expressed genes -  bash version

# ...


#(base) elenagrassi@biotech4neuro:~/biotech4neuro/material/19/solutions$ time snakemake -j 1 expr.bed  expressed_genes_bash.tsv sample_e.geneBodyCoverage.curves.pdf
# very fast
# 25M total
