---
title: "19_20 RNAseq minimal pipeline"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## RNAseq pipeline with Snakemake and R code

For our last laboratory we'll set up a RNASeq pipeline, from fastq up to counts and with some
coverage plots, using Snakemake.

We aim at writing a Snakefile that will automatically follow the steps described by Prof. Molineris in
the RNAseq laboratory in the Bioinformatics module:

https://elearning.unito.it/medicina/mod/url/view.php?id=178796
https://bit.ly/3Dz0IeI

I've created 5 synthetic fastq files to represent 5 samples of RNAsequencing obtained
with an experiment on *Drosophila Melanogaster*, they
are called 'a','b','c','d' and 'e' and you'll find them in '/home/elenagrassi/biotech4neuro/material/19'.
In there you'll also find a skeleton for the Snakefile that you'll need to fill in where you see `# ...`

The suggestion is to review the RNAseq laboratory and divide this laboratory in 4 parts, described below.
They can be implemented separately.

A good way to organize the work could be to decide which step to code, then:

- try it out on the shell manually, to annotate the name of the output file(s) created by the command
- write the rule that will be able to execute it using Snakemake, using the appropriate wildcards to be able to work on all samples
- call it for a single sample to check that it works
- write the 'all' rule to automatically call it for all samples (not needed for all the rules, since for example all
the alignments will be called when you require all the outputs of the count rule, if you chained the rules correctly)

Do not worry about trying out the commands: the synthetic fastq files are very small! You can always use `snakemake -pn`
if you want to check which code would be run with a rule. Just limit yourselves to a single thread for multithreaded
commands!

The Snakefile skeleton has set up some variables with the relevant paths (pointing to STAR indexes, gtf files, ...) - you should
copy the Snakefile in a directory in your home and work there, filling the missing parts.
fastq files, STAR indexes etc should stay where they are (my home or Ivan's), they all are readable
by your users so you will be able to access them as input files in your rules.

Before starting to work remember to use the command `conda` to have access to the commands needed for the
RNAseq pipeline!

### QC

We'll limit ourselves to fastqc - write a rule to call it and an 'all' rule for all samples.
You'll need to point to the directory in my home to use the fastq there (do not copy them to your home to avoid
filling the disk too much!).
Copy the resulting html files to your local computer and inspect them, why do you think the results are so strange?

### Align and count

Use `STAR` and `featureCounts` just like you did during the Bioinformatic lab.
For `featureCounts` use all the parameters suggested by Ivan, pay specific attention to using `-g gene_name`, to have the correct gene
identifiers (needed for the following pipeline steps) in its output.
Write an `all_counts` rule that gets all the output files generated by `featureCounts` and merge
them in a single one (`all_counts.tsv`) with different samples counts as columns - you can do this writing a small R script 
that you'll call in the rule.
To develop the R script interactively you can assign manually to a variable the values that snakemake will put
in the `snakemake@input` for the rule you've written, then try it out in the Rstudio console as always.
When you are satisfied with what your code does, just link this variable to the input and set up the output one, then try out your rule.

Experiment rule chaining: is STAR executed automatically when you require a count file if the bam is not there?
If it's already there? Try different combinations removing the obtained bam/count files.

### Find the expressed genes and obtain the bed with their exons.

If you inspect the tsv that you obtained in the previous section you'll see that the majority of the
genes have 0 reads. Write another rule that calls a R script that'll produce a tsv listing only read counts
for the genes with at least 1 read in at least one sample.

To study the coverage on these expressed genes you'll need to get their exons coordinates from the bed
file which path is in the `BED` variable. But...the identifiers in that bed file are not the same
used by `featureCounts` to count reads :(
You'll need to obtain a file that lists the expressed genes with the same identifiers of the bed.
I've created a file that can act as a *dictionary* for this translation: its path is in the MAP variable.
Write a rule that translates the identifiers of the expressed genes using this map and writes them in a 
column of the file `expressed_genes_counts_flybase.tsv`. Then write another one that starts from BED and this new file 
to create a smaller bed listing only the ids of the expressed genes. 
Try to write the rule for this last step using bash tools instead than writing a
small R script!

### Genebody coverage

Write the rules needed to call `geneBody_coverage.py`. Call it for sample 'a' and 'e', then copy the resulting
plots to your local computer and inspect them.

### Redo everything with a single command!

Experiment how Snakefile behaves: add a new sample to your samples (yes, you guessed right: 'f', the fastq file
is available in the same dir as the others) and call the rule to produce all counts. Which commands are executed? Why?

What would be called if one of the original fastq files would have been changed, instead?

### Optional exercises/homeworks

Extracting the list of expressed genes is doable in bash only, but in this case it would have been
easier to work on single samples (to select only expressed genes) and then put together the info afterwards: try to write the rules
to do this!
(Obtaining `all_counts.tsv` with bash only is doable too, but we did not have time to study all the required tools).